enhancements/machine-api/machine-health-check-pause-on-upgrade.md

```diff
@@ -0,0 +1,175 @@
+      type: Progressing
+    - lastTransitionTime: "2021-07-05T20:05:15Z"
+      message: 'Unable to retrieve available updates: currently reconciling cluster
+        version 4.9.0-0.nightly-2021-07-05-092650 not found in the "stable-4.8" channel'
+      reason: VersionNotFound
+      status: "False"
+      type: RetrievedUpdates
+    ...
+```
+
+The machineHealthCheck supports pausing by setting an annotation on it:
+
+```yaml
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineHealthCheck
+metadata:
+  name: example
+  namespace: openshift-machine-api
+  annotations:
+    cluster.x-k8s.io/paused: ""
+  ...
+```
+
+The new controller will add the pause annotation on the machineHealthChecks when the clusterVersion's `Progressing`
+condition switches to `"True"`. When that condition switches back to `"False"`, the controller will remove the
+annotation.
+
+The new controller will be implemented in the `openshift/machine-api-operator` repository, and be deployed in the
+`machine-api-controllers` pod, both alongside the machine-healthcheck-controller.
+
+### Risks and Mitigations
+
+Nothing we are aware of today
+
+## Design Details
+
+### Open Questions [optional]
+
+1. Should admins be able to enable / disable this feature? (IMHO no, I don't see a usecase when this has disadvantages)
+
+2. Today there is no way to coordinate multiple users of the pause feature of machineHealthChecks. The value of the
+annotation is supposed to be empty. This means that the annotation might already be set by someone else when the new
+controller want to set it as well. That introduces the question what to do in this case after cluster upgrade: remove
+the annotation, or keep it? In the latter case: how to keep track whether the controller set the annotation or not (e.g.
+in another annotation?)
+
+### Test Plan
+
+TODO (need to check existing e2e tests, and if we extend them with a test case for this, and how much extra load to CI
+it would add)
+
+### Graduation Criteria
+
+TODO I have no good idea yet
+- if this is alpha, beta or stable (do we need this at all, it doesn't introduce a new API)?
+- if this is Dev Preview, Tech Preview or GA?
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+#### Removing a deprecated feature
+
+### Upgrade / Downgrade Strategy
+
+The new controller lives in the machine-api-operator image so the upgrades will be driven by the CVO which will fetch
+the right image version as usual.
+The controller does not offer any API or configuration. The controller can be replaced with an older or newer
+version without side effects or risks.
+
+### Version Skew Strategy
+
+This controller has some expectation to other APIs, which are the clusterVersion status conditions, and the
+machineHealthCheck pause annotation. In case any of those changes their API (new condition name or annotaton key), the
+controller will stop to work. This will catched in CI.
+
+## Implementation History
+
+## Drawbacks
+
+## Alternatives
```

Comment:
additional alternative might be a [lease](https://github.com/kubernetes/enhancements/pull/1411).
Before upgrading a node, a lease would need to be acquired.
A lease would also be a prerequisite for remediation.

the main added values are:

1. we don't block all remediations during upgrades, only for nodes being upgraded, and the user gets clear message why remediation is blocked.
2. This could be useful for other players in the system such as Node Maintenance Operator

Reply:
yes, forgot about this one, added, thanks